### PR TITLE
Report loading status for executable dictionaries correctly

### LINKED ISSUE #37572: Dictionaries that failed to execute get reported as successfully loaded
**Describe the unexpected behaviour**
A dictionary that depends on an executable that's missing gets reported as successfully loaded.

**Relevant version**
Observed in 22.3+

**How to reproduce**
Created an integration test for the repro:
[tests/integration/test_dictionaries_broken_config/test.py::test_broken_dict[missing_executable]](https://github.com/ClickHouse/ClickHouse/commit/9d1f8aa684628557a214b16c337ceb27e2fc14cf#diff-aad2ab88d46a3363a74367a643c8978338f5dbdb847864a741119cf7bf4bc584R44)

TLDR is

1. Create a dictionary with the following configuration:
```
  <source>
    <executable>
      <!-- OOPS! executable not found -->
      <command>executable_not_found</command>
      <format>CSV</format>
    </executable>
  </source>
```
2. `SELECT status FROM system.dictionaries` will show `LOADED`

**Expected behavior**
`21.8` behaves differently (more correct IMO):

 `SELECT status FROM system.dictionaries` will show `FAILED`
`dictGet` will result in client receiving an exception `DB::Exception: Child process was exited with return code 127`

**Additional context**
Possibly related issue: https://github.com/ClickHouse/ClickHouse/issues/37571

### PR DESCRIPTION
**Pull request by tonickkozlov**
# Changelog category (leave one):
* Bug Fix (user-visible misbehavior in an official stable release)

# Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Report correct status (FAILED) for executable dictionaries that failed to load.

Fixes #37572

### CODE REVIEW

**Comment by tonickkozlov** [comment]
A real use-case for this change is doing a python2 to python3. If you e.g. start using a base image that only ships with python3, all the dictionaries that rely on python2 will break quietly.

File: src/Common/waitForPid.cpp [hunk 1]
```diff
@@ -157,7 +157,7 @@ static PollPidResult pollPid(pid_t pid, int timeout_in_ms)
     #error "Unsupported OS type"
 #endif
 
-bool waitForPid(pid_t pid, size_t timeout_in_seconds)
+bool waitForPid(pid_t pid, size_t timeout_in_seconds, int* out_status)
```
Comment (evillique) [inline src/Common/waitForPid.cpp:160]:
```suggestion
bool waitForPid(pid_t pid, size_t timeout_in_seconds, int * out_status)
```

File: src/Common/waitForPid.h [hunk 2]
```diff
@@ -7,6 +7,6 @@ namespace DB
  * Waits for a specific pid with timeout
  * Returns `true` if process terminated successfully in specified timeout or `false` otherwise
  */
-bool waitForPid(pid_t pid, size_t timeout_in_seconds);
+bool waitForPid(pid_t pid, size_t timeout_in_seconds, int * out_pid = nullptr);
```
Comment (evillique) [inline src/Common/waitForPid.h:10]:
```suggestion
bool waitForPid(pid_t pid, size_t timeout_in_seconds, int * out_status = nullptr);
```

**Review by evillique — COMMENTED** [review]
Please take a look at integration tests `test_executable_dictionary` and the new `test_dictionaries_broken_config` as they fail in the CI.

**Comment by tonickkozlov** [comment]
Thanks for your review @evillique , updated my PR.

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 5b7b424d634ae614d0ec450526454ee9074f88c2 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48775/5b7b424d634ae614d0ec450526454ee9074f88c2/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by alexey-milovidov** [comment]
@tonickkozlov LGTM, and the tests passed.

But I didn't understand whether we needed a new configuration parameter. Why don't always check?
And if we will end up introducing a new configuration parameter, it should be covered by docs.

**Comment by tonickkozlov** [comment]
@alexey-milovidov Thank you for taking a look. I also think that checking for exit status code should be the default behaviour, no configuration parameter required.
When I tried that previously it broke tests regarding default dictionary column values. My understanding is that the default behaviour is not reporting the executable error but silently filling in the default values. I have just made a change to always check exit statuses so we can verify that's the case.
I'm not sure how to consolidate the two behaviours (reporting the error and filling in the default values). Maybe someone from your team could help? 🙏  Really appreciate it!

**Comment by alexey-milovidov** [comment]
I don't think filling the default values was intended. Maybe change it unconditionally, and it will be alright?

**Comment by tonickkozlov** [comment]
@alexey-milovidov I've simplified the code and updated the integration test to reflect the desired behaviour.
`00534_functions_bad_arguments` stateless test is failing in msan, but I can't see why. Could it be related? Thank you.

**Comment by alexey-milovidov** [comment]
Ok. One question - do we need the "strict_exit_code_check" parameter? If we do, can we rename it to "check_exit_code", because, if I understood correctly, it means "check the exit code".

**Comment by alexey-milovidov** [comment]
@tonickkozlov one question remains.

**Comment by tonickkozlov** [comment]
@alexey-milovidov I think the parameter can be useful to control behaviour more granularly. For example, UDFs also rely on the ShellCommandSource, and I don't want to break the current default behaviour (falling back to default) in case some users depend on it. So, my change affects only dictionaries. Does it make sense?

**Comment by alexey-milovidov** [comment]
Yes, all good.

I'll review the "metadata_loading" test and we will merge.

**Comment by tonickkozlov** [comment]
Thank you for quick turn-around @alexey-milovidov 🙏

**Pull request merged**